addons/mail/sfu_server/src/client.js

```diff
@@ -0,0 +1,574 @@
+            }, RECOVERY_DELAY);
+            return;
+        }
+        this._onCleanup(() => {
+            this._producers[type]?.close();
+            this._producers[type] = null;
+            clearTimeout(this._recoverProducerTimeouts[type]);
+        });
+    }
+
+    /**
+     * To be overridden in tests.
+     *
+     * @returns {import("mediasoup-client").types.Device}
+     */
+    _createDevice() {
+        return new Device();
+    }
+
+    /**
+     * To be overridden in tests.
+     *
+     * @param {string} url
+     * @returns {WebSocket}
+     * @private
+     */
+    _createWebSocket(url) {
+        return new WebSocket(url);
+    }
+
+    /**
+     * Opens the webSocket connection to the server and authenticates, handles reconnection attempts.
+     */
+    async _connect() {
+        if (ACTIVE_STATES.has(this.state)) {
+            return;
+        }
+        this._clear();
+        this.state = SFU_CLIENT_STATE.CONNECTING;
+        try {
+            this._bus = await this._createBus();
+        } catch {
+            this._handleConnectionEnd();
+            return;
+        }
+        this._bus.onMessage = this._handleMessage;
+        this._bus.onRequest = this._handleRequest;
+    }
+
+    /**
+     * @returns {Promise<Bus>}
+     */
+    _createBus() {
+        return new Promise((resolve, reject) => {
+            let webSocket;
+            try {
+                webSocket = this._createWebSocket(this._url);
+            } catch (error) {
+                reject(error);
+                return;
+            }
+            webSocket.addEventListener("close", this._handleConnectionEnd);
+            webSocket.addEventListener("error", this._handleConnectionEnd);
+            this._onCleanup(() => {
+                webSocket.removeEventListener("close", this._handleConnectionEnd);
+                webSocket.removeEventListener("error", this._handleConnectionEnd);
+                webSocket.close(WS_CLOSE_CODE.CLEAN);
+            });
+            // Websocket handshake with the rtc server,
+            // when opening the webSocket, the server expects the first message to contain the jwt,
+            // all subsequent messages should be handled by the Bus.
+            webSocket.addEventListener(
+                "message",
+                () => {
+                    this.state = SFU_CLIENT_STATE.AUTHENTICATED;
+                    resolve(new Bus(webSocket));
+                },
+                { once: true }
+            );
+            webSocket.addEventListener(
+                "open",
+                () => {
+                    webSocket.send(JSON.stringify(this._jsonWebToken));
+                },
+                { once: true }
+            );
+        });
+    }
+
+    /**
+     * @param {Function} callback
+     */
+    _onCleanup(callback) {
+        this._cleanups.push(callback);
+    }
+
+    _clear() {
+        for (const cleanup of this._cleanups.slice(0)) {
```

Comment:
this needs to be `splice` unless you want the cleanups to stay for the next call to clear. Alternatively you can also just set the cleanups to a new empty array after the loop